### DISCUSSION TITLE
HtmlRenderer seems to ignore setting GENERATE_HEADER_ID in some cases

### DISCUSSION THREAD

**Discussion by travkin79** [vsch/flexmark-java#565] [Q&A]
I am not sure if I observe a bug or if it is desired behavior.

I'm using the Markdown parser and HTML renderer with the following settings (besides others):

- GENERATE_HEADER_ID=false
- RENDER_HEADER_ID=true
- AttributesExtension activated

**I expect:**
If I explicitly set header IDs in headings, they will be rendered, but no IDs will be generated for headings that do not have any headings ID in Markdown source code. For example:

```
# Heading 1

## Images {#section-images}

```

should become

```
<h1>Heading 1</h1>
<h2 id="section-images">Images</h2>
```

**Instead I see:**
If I explicitly set header IDs in headings, they will be rendered. If they are not set, they will be generated and rendered, too. For the example from above, we'll get:

```
<h1 id="heading-1">Heading 1</h1>
<h2 id="section-images">Images</h2>
```

Do I misunderstand something or is that a bug?

I traced down the source code to the following statement in `com.vladsch.flexmark.html.HtmlRenderer#MainNodeRenderer`:

```
this.htmlIdGenerator = htmlIdGeneratorFactory != null ? htmlIdGeneratorFactory.create(this)
                    : (!(htmlOptions.renderHeaderId || htmlOptions.generateHeaderIds) ? HtmlIdGenerator.NULL : new HeaderIdGenerator.Factory().create(this));
```

I think, an ID generator should only be created if the `generateHeaderIds` settings is active and should be independent from the renderHeaderId setting. Do you agree?

**Comment by vsch** [comment]
@travkin79, it is a left-over from legacy implementation. You are right the `htmlOptions.renderHeaderId` should not be used in the test.

I will be addressing all the accumulated PR's and bugs in the next couple of weeks.

**Comment by travkin79** [comment]
Thank you for the quick answer, @vsch. I've prepared a pull request #566 to fix that.

**Comment by travkin79** [comment]
Hi @vsch,
There still seems to be a problem or I misunderstand something.

I'm using flexmark 0.64.6 now and I'm having

- GENERATE_HEADER_ID=false
- RENDER_HEADER_ID=true
- AttributesExtension activated

I have a use case, where I want to render something like
`## Images {#section-images}`

When debugging, I see

- the `HtmlRenderer` uses an `HtmlIdGenerator` in `HtmlRenderer#getNodeId(Node)`
- the `HtmlIdGenerator` does return `null` (I think, because I have `GENERATE_HEADER_ID=false`)
- the current `Node`'s `anchorRefId` attribute is correctly set to `section-images` what I expected to be the ID to be rendered
- the `HtmlRenderer#getNodeId(Node)` method does not read the `Node`'s `anchorRefId` attribute and returns `null` instead
- the `CoreNodeRenderer#render(Heading, NodeRendererContext, HtmlWriter)` method doesn't read the `Node`'s `anchorRefId` attribute either

Shouldn't the `Node`'s `anchorRefId` attribute be used as an ID in this case?

The final result is something I wouldn't expect with these settings, a heading without any ID:
`<h2>Images</h2>`

By the way, I also tried using flexmark 0.64.8, but I couldn't read the `flexmark-java-0.64.8.zip` sources file. It seems to be corrupt.

Best regards,
Dietrich

**Comment by travkin79** [comment]
Hi @vsch,
I found the time to create a pull request with a unit test to reproduce my situation and with a bug fix (see pull request #581). Could you please review and merge the bug fix?

Thank you and best regards.
Dietrich

**Reply by travkin79** [reply]
Hi @vsch,
Is there any update on that?
Best regards.
Dietrich
